src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.server.permission.events;
+
+import com.google.common.base.Preconditions;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.server.permission.handler.IPermissionHandler;
+import net.minecraftforge.server.permission.nodes.PermissionNode;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * <p>Fired during Server startup<br/>
+ * {@link Handler} allows to set a new PermissionHandler<br/>
+ * {@link Nodes} allows you to register new PermissionNodes</p>
+ *
+ * <p><strong>Note:</strong> All PermissionNodes that you want to use, <strong>must</strong> be registered!</p>
+ */
+public class PermissionGatherEvent extends Event
+{
+
+   /**
+    * Used to register a new PermissionHandler, last set one is chosen
+    */
+   public static class Handler extends PermissionGatherEvent
+   {
+      private IPermissionHandler currentHandler;
+
+      public Handler(IPermissionHandler currentHandler)
+      {
+         this.currentHandler = currentHandler;
+      }
+
+      public IPermissionHandler getCurrentHandler()
+      {
+         return currentHandler;
+      }
+
+      public void setCurrentHandler(IPermissionHandler newHandler)
+      {
+         Preconditions.checkNotNull(newHandler, "Permission handler cannot be null!");
+         this.currentHandler = newHandler;
+      }
+   }
+
+
+   /**
+    * Used to register your PermissionNodes, <strong>every node that you want to use, must be registered!</strong>
+    */
+   public static class Nodes extends PermissionGatherEvent
+   {
+      private Set<PermissionNode<?>> nodes;
+
+      public Nodes()
+      {
+         this.nodes = new HashSet<>();
+      }
+
+      public Set<PermissionNode<?>> getNodes()
+      {
+         return nodes;
+      }
+
+      public void addNodes(PermissionNode<?>... nodes)
```

Comment:
What about two nodes with the same `nodeName`?

Reply:
Do Hashset work in Buckets? If not they'll be removed. As they have the same hashes.

Reply:
Maybe it should be considered to explicitly test if the node is already listed in this event, and throw an exception if so?